// slonik/dist/connectionMethods/query.js
import { executeQuery } from '../routines/executeQuery.js';
import { SlonikError } from '@slonik/errors';
const executionRoutine = async (finalConnection, finalSql, finalValues) => {
    const result = await finalConnection.query(finalSql, 
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    finalValues);
    const fields = [];
    if (result.fields) {
        for (const field of result.fields) {
            fields.push({
                dataTypeId: field.dataTypeId,
                name: field.name,
            });
        }
    }
    return {
        command: result.command,
        fields,
        notices: result.notices ?? [],
        rowCount: result.rowCount || 0,
        rows: (result.rows || []),
        type: 'QueryResult',
    };
};
export const query = async (connectionLogger, connection, clientConfiguration, slonikSql, inheritedQueryId, integrityValidation) => {
    try {
        return await executeQuery(connectionLogger, connection, clientConfiguration, slonikSql, inheritedQueryId, executionRoutine, false, integrityValidation);
    }
    catch (error) {
        if (error instanceof SlonikError) {
            connection.events.emit('error', error);
        }
        throw error;
    }
};
