// slonik/dist/test/slonik/routines/executeQuery.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ava_1 = __importDefault(require("ava"));
const roarr_1 = require("roarr");
const sinon = __importStar(require("sinon"));
const errors_1 = require("../../../src/errors");
const executeQuery_1 = require("../../../src/routines/executeQuery");
const state_1 = require("../../../src/state");
const createClientConfiguration_1 = require("../../helpers/createClientConfiguration");
const createErrorWithCode_1 = require("../../helpers/createErrorWithCode");
const test = ava_1.default;
const { beforeEach, } = test;
const createConnectionStub = () => {
    return {
        connection: {
            slonik: {
                terminated: null,
            },
        },
        off() { },
        on() { },
    };
};
beforeEach((t) => {
    t.context.logger = roarr_1.Roarr;
    t.context.connection = createConnectionStub();
    t.context.executionRoutine = () => { };
    state_1.poolClientStateMap.set(t.context.connection, {
        connectionId: '1',
        mock: true,
        poolId: '1',
        terminated: null,
        transactionDepth: null,
        transactionId: null,
    });
});
test('throws a descriptive error if query is empty', async (t) => {
    const error = await t.throwsAsync(async () => {
        return await (0, executeQuery_1.executeQuery)(t.context.logger, t.context.connection, (0, createClientConfiguration_1.createClientConfiguration)(), {
            sql: '',
            values: [],
        }, 'foo', t.context.executionRoutine);
    });
    t.true(error instanceof errors_1.InvalidInputError);
    t.is(error === null || error === void 0 ? void 0 : error.message, 'Unexpected SQL input. Query cannot be empty.');
});
test('throws a descriptive error if the entire query is a value binding', async (t) => {
    const error = await t.throwsAsync(async () => {
        return await (0, executeQuery_1.executeQuery)(t.context.logger, t.context.connection, (0, createClientConfiguration_1.createClientConfiguration)(), {
            sql: '$1',
            values: [],
        }, 'foo', t.context.executionRoutine);
    });
    t.true(error instanceof errors_1.InvalidInputError);
    t.is(error === null || error === void 0 ? void 0 : error.message, 'Unexpected SQL input. Query cannot be empty. Found only value binding.');
});
test('retries an implicit query that failed due to a transaction error', async (t) => {
    const executionRoutineStub = sinon.stub();
    executionRoutineStub.onFirstCall()
        .rejects((0, createErrorWithCode_1.createErrorWithCode)('40P01'))
        .onSecondCall()
        .resolves({
        command: 'SELECT',
        fields: [],
        notices: [],
        rowCount: 1,
        rows: [
            {
                foo: 1,
            },
        ],
    });
    const result = await (0, executeQuery_1.executeQuery)(t.context.logger, t.context.connection, (0, createClientConfiguration_1.createClientConfiguration)(), {
        sql: 'SELECT 1 AS foo',
        values: [],
    }, 'foo', executionRoutineStub);
    t.is(executionRoutineStub.callCount, 2);
    t.deepEqual(result, {
        command: 'SELECT',
        fields: [],
        notices: [],
        rowCount: 1,
        rows: [
            {
                foo: 1,
            },
        ],
    });
});
test('returns the thrown transaction error if the retry limit is reached', async (t) => {
    const executionRoutineStub = sinon.stub();
    executionRoutineStub.onFirstCall()
        .rejects((0, createErrorWithCode_1.createErrorWithCode)('40P01'))
        .onSecondCall()
        .rejects((0, createErrorWithCode_1.createErrorWithCode)('40P01'));
    const clientConfiguration = (0, createClientConfiguration_1.createClientConfiguration)();
    const error = await t.throwsAsync((0, executeQuery_1.executeQuery)(t.context.logger, t.context.connection, {
        ...clientConfiguration,
        queryRetryLimit: 1,
    }, {
        sql: 'SELECT 1 AS foo',
        values: [],
    }, 'foo', executionRoutineStub));
    t.is(executionRoutineStub.callCount, 2);
    t.true(error instanceof Error);
    t.is(error.code, '40P01');
});
test('transaction errors are not handled if the function was called by a transaction', async (t) => {
    const connection = createConnectionStub();
    state_1.poolClientStateMap.set(connection, {
        connectionId: '1',
        mock: true,
        poolId: '1',
        terminated: null,
        transactionDepth: null,
        transactionId: '1',
    });
    const executionRoutineStub = sinon.stub();
    executionRoutineStub.onFirstCall()
        .rejects((0, createErrorWithCode_1.createErrorWithCode)('40P01'));
    const clientConfiguration = (0, createClientConfiguration_1.createClientConfiguration)();
    const error = await t.throwsAsync((0, executeQuery_1.executeQuery)(t.context.logger, connection, {
        ...clientConfiguration,
        queryRetryLimit: 1,
    }, {
        sql: 'SELECT 1 AS foo',
        values: [],
    }, 'foo', executionRoutineStub));
    t.is(executionRoutineStub.callCount, 1);
    t.true(error instanceof Error);
    t.is(error.code, '40P01');
});
