// slonik/dist/test/types.js
"use strict";
/**
 * Functions in this file are never actually run - they are purely
 * a type-level tests to ensure the typings don't regress.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.queryMethods = void 0;
const expect_type_1 = require("expect-type");
const zod_1 = require("zod");
const src_1 = require("../src");
const ZodRow = zod_1.z.object({
    bar: zod_1.z.boolean(),
    foo: zod_1.z.string(),
});
const queryMethods = async () => {
    const client = await (0, src_1.createPool)('');
    const sql = (0, src_1.createSqlTag)({
        typeAliases: {
            row: ZodRow,
        },
    });
    // parser
    const parser = sql.type(ZodRow) ``.parser;
    (0, expect_type_1.expectTypeOf)(parser).toEqualTypeOf();
    // any
    const any = await client.any(sql ``);
    (0, expect_type_1.expectTypeOf)(any).toEqualTypeOf();
    const anyTyped = await client.any(sql ``);
    (0, expect_type_1.expectTypeOf)(anyTyped).toEqualTypeOf();
    const anyTypedQuery = await client.any(sql ``);
    (0, expect_type_1.expectTypeOf)(anyTypedQuery).toEqualTypeOf();
    const anyZodTypedQuery = await client.any(sql.type(ZodRow) ``);
    (0, expect_type_1.expectTypeOf)(anyZodTypedQuery).toEqualTypeOf();
    const anyZodAliasQuery = await client.any(sql.typeAlias('row') ``);
    (0, expect_type_1.expectTypeOf)(anyZodAliasQuery).toEqualTypeOf();
    // anyFirst
    const anyFirst = await client.anyFirst(sql ``);
    (0, expect_type_1.expectTypeOf)(anyFirst).toEqualTypeOf();
    const anyFirstTyped = await client.anyFirst(sql ``);
    (0, expect_type_1.expectTypeOf)(anyFirstTyped).toEqualTypeOf();
    const anyFirstTypedQuery = await client.anyFirst(sql ``);
    (0, expect_type_1.expectTypeOf)(anyFirstTypedQuery).toEqualTypeOf();
    const anyFirstZodTypedQuery = await client.anyFirst(sql.type(ZodRow) ``);
    (0, expect_type_1.expectTypeOf)(anyFirstZodTypedQuery).toEqualTypeOf();
    // many
    const many = await client.many(sql ``);
    (0, expect_type_1.expectTypeOf)(many).toEqualTypeOf();
    const manyTyped = await client.many(sql ``);
    (0, expect_type_1.expectTypeOf)(manyTyped).toEqualTypeOf();
    const manyTypedQuery = await client.many(sql ``);
    (0, expect_type_1.expectTypeOf)(manyTypedQuery).toEqualTypeOf();
    const manyZodTypedQuery = await client.many(sql.type(ZodRow) ``);
    (0, expect_type_1.expectTypeOf)(manyZodTypedQuery).toEqualTypeOf();
    // manyFirst
    const manyFirst = await client.manyFirst(sql ``);
    (0, expect_type_1.expectTypeOf)(manyFirst).toEqualTypeOf();
    const manyFirstTyped = await client.manyFirst(sql ``);
    (0, expect_type_1.expectTypeOf)(manyFirstTyped).toEqualTypeOf();
    const manyFirstTypedQuery = await client.manyFirst(sql ``);
    (0, expect_type_1.expectTypeOf)(manyFirstTypedQuery).toEqualTypeOf();
    const manyFirstZodTypedQuery = await client.manyFirst(sql.type(ZodRow) ``);
    (0, expect_type_1.expectTypeOf)(manyFirstZodTypedQuery).toEqualTypeOf();
    // maybeOne
    const maybeOne = await client.maybeOne(sql ``);
    (0, expect_type_1.expectTypeOf)(maybeOne).toEqualTypeOf();
    const maybeOneTyped = await client.maybeOne(sql ``);
    (0, expect_type_1.expectTypeOf)(maybeOneTyped).toEqualTypeOf();
    const maybeOneTypedQuery = await client.maybeOne(sql ``);
    (0, expect_type_1.expectTypeOf)(maybeOneTypedQuery).toEqualTypeOf();
    const maybeOneZodTypedQuery = await client.maybeOne(sql.type(ZodRow) ``);
    (0, expect_type_1.expectTypeOf)(maybeOneZodTypedQuery).toEqualTypeOf();
    // maybeOneFirst
    const maybeOneFirst = await client.maybeOneFirst(sql ``);
    (0, expect_type_1.expectTypeOf)(maybeOneFirst).toEqualTypeOf();
    const maybeOneFirstTyped = await client.maybeOneFirst(sql ``);
    (0, expect_type_1.expectTypeOf)(maybeOneFirstTyped).toEqualTypeOf();
    const maybeOneFirstTypedQuery = await client.maybeOneFirst(sql ``);
    (0, expect_type_1.expectTypeOf)(maybeOneFirstTypedQuery).toEqualTypeOf();
    const maybeOneFirstZodTypedQuery = await client.maybeOneFirst(sql.type(ZodRow) ``);
    (0, expect_type_1.expectTypeOf)(maybeOneFirstZodTypedQuery).toEqualTypeOf();
    // one
    const one = await client.one(sql ``);
    (0, expect_type_1.expectTypeOf)(one).toEqualTypeOf();
    const oneTyped = await client.one(sql ``);
    (0, expect_type_1.expectTypeOf)(oneTyped).toEqualTypeOf();
    const oneTypedQuery = await client.one(sql ``);
    (0, expect_type_1.expectTypeOf)(oneTypedQuery).toEqualTypeOf();
    const oneZodTypedQuery = await client.one(sql.type(ZodRow) ``);
    (0, expect_type_1.expectTypeOf)(oneZodTypedQuery).toEqualTypeOf();
    // oneFirst
    const oneFirst = await client.oneFirst(sql ``);
    (0, expect_type_1.expectTypeOf)(oneFirst).toEqualTypeOf();
    const oneFirstTyped = await client.oneFirst(sql ``);
    (0, expect_type_1.expectTypeOf)(oneFirstTyped).toEqualTypeOf();
    const oneFirstTypedQuery = await client.oneFirst(sql ``);
    (0, expect_type_1.expectTypeOf)(oneFirstTypedQuery).toEqualTypeOf();
    const oneFirstZodTypedQuery = await client.oneFirst(sql.type(ZodRow) ``);
    (0, expect_type_1.expectTypeOf)(oneFirstZodTypedQuery).toEqualTypeOf();
    // query
    const query = await client.query(sql ``);
    (0, expect_type_1.expectTypeOf)(query).toMatchTypeOf();
    const queryTyped = await client.query(sql ``);
    (0, expect_type_1.expectTypeOf)(queryTyped).toMatchTypeOf();
    const queryTypedQuery = await client.query(sql ``);
    (0, expect_type_1.expectTypeOf)(queryTypedQuery).toMatchTypeOf();
    const queryZodTypedQuery = await client.query(sql.type(ZodRow) ``);
    (0, expect_type_1.expectTypeOf)(queryZodTypedQuery).toMatchTypeOf();
    const FooBarRow = zod_1.z.object({
        x: zod_1.z.string(),
        y: zod_1.z.number(),
    });
    (0, expect_type_1.expectTypeOf)(await client.one(sql.type(FooBarRow) `select 'x' x, 123 y`)).toEqualTypeOf();
    (0, expect_type_1.expectTypeOf)(await client.any(sql.type(FooBarRow) `select 'x' x, 123 y`)).toEqualTypeOf();
};
exports.queryMethods = queryMethods;
