// slonik/dist/test/helpers/createPool.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPool = void 0;
const events_1 = __importDefault(require("events"));
const sinon = __importStar(require("sinon"));
const bindPool_1 = require("../../src/binders/bindPool");
const state_1 = require("../../src/state");
const Logger_1 = require("./Logger");
const defaultConfiguration = {
    interceptors: [],
    typeParsers: [],
};
const createPool = async (clientConfiguration = defaultConfiguration) => {
    const eventEmitter = new events_1.default();
    const connection = {
        connection: {
            slonik: {
                connectionId: '1',
                mock: false,
                poolId: '1',
                transactionDepth: null,
            },
        },
        emit: eventEmitter.emit.bind(eventEmitter),
        end: () => { },
        off: eventEmitter.off.bind(eventEmitter),
        on: eventEmitter.on.bind(eventEmitter),
        query: () => {
            return {};
        },
        release: () => { },
    };
    const internalPool = {
        _pulseQueue: () => { },
        _remove: () => { },
        connect: () => {
            return connection;
        },
    };
    state_1.poolStateMap.set(internalPool, {
        ended: false,
        mock: false,
        poolId: '1',
        typeOverrides: null,
    });
    const connectSpy = sinon.spy(internalPool, 'connect');
    const endSpy = sinon.spy(connection, 'end');
    const querySpy = sinon.stub(connection, 'query').returns({});
    const releaseSpy = sinon.spy(connection, 'release');
    const removeSpy = sinon.spy(internalPool, '_remove');
    const pool = (0, bindPool_1.bindPool)(Logger_1.Logger, internalPool, {
        captureStackTrace: false,
        connectionRetryLimit: 1,
        connectionTimeout: 5000,
        idleInTransactionSessionTimeout: 5000,
        idleTimeout: 5000,
        interceptors: [],
        maximumPoolSize: 1,
        queryRetryLimit: 1,
        statementTimeout: 5000,
        transactionRetryLimit: 1,
        typeParsers: [],
        ...clientConfiguration,
    });
    return {
        ...pool,
        connection,
        connectSpy,
        endSpy,
        querySpy,
        releaseSpy,
        removeSpy,
    };
};
exports.createPool = createPool;
